src/main/java/org/sagebionetworks/web/client/widget/ReactComponentV2.java

```diff
@@ -0,0 +1,277 @@
+ * <p>
+ * This widget automatically unmounts the ReactComponent (if any) when this container is detached/unloaded.
+ */
+public abstract class ReactComponentV2<
+  T extends ReactComponentType<P>, P extends ReactComponentProps
+>
+  extends ComplexPanel
+  implements HasClickHandlers {
+
+  private ReactDOMRoot root;
+  private final ReactComponentType<P> reactComponentType;
+  public P props;
+
+  private final ArrayList<Element> nonReactChildElements = new ArrayList<>();
+
+  public ReactComponentV2(T reactComponentType, P props) {
+    this(reactComponentType, props, DivElement.TAG);
+  }
+
+  public ReactComponentV2(T reactComponentType, P props, String tag) {
+    this.reactComponentType = reactComponentType;
+    this.props = props;
+    setElement(Document.get().createElement(tag));
+  }
+
+  /**
+   * If any children of this widget do not implement this class, then we will inject them into the DOM using the React
+   * component's `ref` prop. This method will update the props with a callback ref that handles appending the children
+   * to the DOM node on which the ref is forwarded.
+   */
+  private void maybeUpdatePropsWithCallbackRef() {
+    if (!this.allChildrenAreReactComponents() && getChildren().size() > 0) {
+      // Create a callback ref that will allow us to inject the GWT children into the DOM
+      ReactComponentProps.CallbackRef callbackRef = (Element node) -> {
+        if (node != null) {
+          // Once the DOM node is defined, inject each child
+          getChildren()
+            .forEach(w -> {
+              node.appendChild(w.getElement());
+              // Keep track of child elements to ensure they are removed when the component unloads or re-renders
+              nonReactChildElements.add(w.getElement());
+            });
+        }
+      };
+
+      if (this.props == null) {
+        this.props = (P) JsPropertyMap.of();
+      }
+      // Override the ref
+      this.props.ref = callbackRef;
+    }
+  }
+
+  private void createRoot() {
+    if (root == null) {
+      root = ReactDOM.createRoot(this.getElement());
+    }
+  }
+
+  private void destroyRoot() {
+    if (root != null) {
+      // Asynchronously schedule unmounting the root to allow React to finish the current render cycle.
+      // https://github.com/facebook/react/issues/25675
+      Timer t = new Timer() {
+        @Override
+        public void run() {
+          root.unmount();
+          root = null;
+        }
+      };
+      t.schedule(0);
+    }
+  }
+
+  private void detachNonReactChildElements() {
+    if (allChildrenAreReactComponents()) {
+      // No need to remove non-React child elements from this widget
+      // But a descendant contain non-React children, so recurse!
+      for (Widget w : getChildren()) {
+        ((ReactComponentV2<?, ?>) w).detachNonReactChildElements();
+      }
+    } else {
+      nonReactChildElements.forEach(Element::removeFromParent);
+      nonReactChildElements.clear();
+    }
```

Comment:
Recursively check descendants for non-GWT elements to detach until we've checked the entire widget/component tree